### ISSUE TITLE
DAX unit testing

### ISSUE THREAD

**Issue by atbenmurray** [VUIIS/dax#164]
Issue to track refactoring to make more functionality in DAX unit testable.

**Comment by damonms** [comment]
I've been out of the loop a bit here for a while now, but my suggestion would be to first run integration testing. Unit testing is better done bottom up and that would be quite an undertaking considering the extensive moving targets here.

More specifically, I would suggest up booting up several commponents:

1. A virtualized gird  (slurm, torque/moab, sge etc)
2. A virtualized HTTP XNAT instance
3. A virtualized HTTPS XNAT instance

Ideally these could all be docker instances so the entire system could be spun up and torn down as needed before releases.

**Comment by atbenmurray** [comment]
Hi Damon,
I've just done a major refactoring of some of the processor internals to make yaml processor definitions much more flexible. In order to do that, I've written some tests that, with a little bit of cleaning up, test a good amount of XnatUtils, Processor and related items. I'd like to make them the starting point of a Jenkins CI setup. This includes limited mocks of the pyxnat-derived InterfaceTemp class, XnatUtils itself (although I've since moved some of that functionality into InterfaceTemp on the branch 'asr_of_asr') and the CachedImage* classes.
At the integration testing end of the scale, I've started putting together an integration test designed to run with a real xnat and test-specific processors/spiders/work-queues. So far its all limited to what I needed to sanity check the extensive changes that I made. It all needs (limited) further work to be functional but after that they could also go into a CI/CD setup.
Happy to discuss more.
Ben
